applications/FluidDynamicsApplication/custom_utilities/mass_conservation_utility.cpp

```diff
@@ -30,131 +29,211 @@ namespace Kratos
+        // case: water volume was gained by mistake
+        mFluidVolumeConservation = FluidVolumeConservation::VOLUME_GAINED;
+        double water_inflow_over_boundary = OrthogonalFlowIntoAir( -1.0 );
+        if ( r_comm.GetDataCommunicator().SumAll( water_inflow_over_boundary ) ){
+            // checking if flow is sufficient (avoid division by 0)
+            if ( water_inflow_over_boundary > 1.0e-12 ){
+                time_step_for_convection = - mWaterVolumeError / water_inflow_over_boundary;
+            }
+        } else {
+            KRATOS_ERROR << "Communication failed in MassConservationUtility::ComputeDtForConvection()";
+        }
+    }
+    else {
+        // case: Exactly the correct volume of water is present
+        mFluidVolumeConservation = FluidVolumeConservation::EXACT_VOLUME;
+    }
+
+    if (std::abs(time_step_for_convection) > mMaxDt) {
+        time_step_for_convection = time_step_for_convection > 0 ? mMaxDt : -mMaxDt;
+    } else if (std::abs(time_step_for_convection) < mMinDt) {
+        time_step_for_convection = time_step_for_convection > 0 ? mMinDt : -mMinDt;
+    }
+
+    KRATOS_WARNING_IF("MassConservationUtility", time_step_for_convection < 0.0) << "A time step smaller than 0.0 was computed." << std::endl;
+
+    return time_step_for_convection;
+}
+
+
+
+void MassConservationUtility::ApplyLocalCorrection( const Variable<double>& rAuxDistVar ){
+
+    const auto node_begin = mrModelPart.GetCommunicator().LocalMesh().NodesBegin();
+    const int number_nodes = mrModelPart.GetCommunicator().LocalMesh().NumberOfNodes();
+
+    #pragma omp parallel for
+    for (int i_node = 0; i_node < number_nodes; ++i_node){
+        auto it_node = node_begin + i_node;
+        double& r_original_dist = it_node->FastGetSolutionStepValue( DISTANCE, 0 );
+        const double& r_aux_dist = it_node->GetValue( rAuxDistVar );
+
+        if ( mFluidVolumeConservation == FluidVolumeConservation::VOLUME_LOST ){
+            // choosing minimum to extend water domain
+            r_original_dist = std::min( r_original_dist, r_aux_dist );
+        } else if (mFluidVolumeConservation == FluidVolumeConservation::VOLUME_GAINED ){
+            // choosing maximum to reduce water domain
+            r_original_dist = std::max( r_original_dist, r_aux_dist );
+        }
+    }
+}
+
+
+
+void MassConservationUtility::ReCheckTheMassConservation(){
+
+    // recomputation based on the new distance field.
+    double pos_vol = 0.0;
+    double neg_vol = 0.0;
+    double inter_area = 0.0;
+    ComputeVolumesAndInterface( pos_vol, neg_vol, inter_area );
+
+    mWaterVolumeError = mTheoreticalNegativeVolume - neg_vol;
+    mInterfaceArea = inter_area;
+    KRATOS_INFO_IF("MassConservationUtility", mEchoLevel > 0) << "Volume error after correction: " << mWaterVolumeError << std::endl;
+}
+
+
+
+void MassConservationUtility::ApplyGlobalCorrection(){
+
+    const double inter_area = mInterfaceArea;
```

Comment:
Why do we need `inter_area`? I think we can directly operate with `mInterfaceArea`.